GUI Working in NWN2Dev

scottmunday84 opened the pull request:
You will have to install VCPKG and install the following debug/release packages for the triplet x86-windows-static-md (soon to be changed to x86-windows-static, but required the loose bind to the CRT engine, since that is how the original code was setup):

* Detours
* wxWidgets

These dependencies is (clearly) too much for the git repository. Eventually, I will build a script to perform this action. Otherwise, my code now works with your code using the latest version of wxWidgets and MS Detours. 

Here's a few of my other notes:
* I made Log() and LogStr() virtual on your new LogNWNX class. This is so I can log entries to the wxWidgets application while still leveraging your new awesome logger.
* The system now builds with Unicode. No ATL conversions.
* The work is now built through a CMake build. Your work and mine merged in pretty smoothly.
* I kept the hook_horror logo in the code. I will be using it when I log xp_rpc extension I'm developing.
* wxWidgets was dropped from all areas outside of the GUI project.
* Dropped all gRPC references; will place in another branch.

Regards.

scottmunday84 commented:
![image](https://user-images.githubusercontent.com/925641/119293097-b9137400-bc1f-11eb-8e48-5342eb06c550.png)

scottmunday84 commented:
I also want to explain a little more on something I'll be working on.

With the gRPC, it allows you a way to create services and their clients using protobufs to define both services and message structure. The first test I'm going to try to build gRPC with an IPC communication using winsock. This will allow us to communicate directly to the controller (which knows the logger instance used), and log either in the application, the console, or a file.

SkywingvL commented on src/hook/hook.cpp:
```diff
@@ -772,6 +774,8 @@ int WINAPI NWNXWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,
 
 	shmem = nullptr;
 
+	std::this_thread::sleep_for(10s);
```
This seems potentially questionable; what are we delaying for?

scottmunday84 replied:
This is all a WIP. I placed it in there so I could space out my debugging of the shared memory object. I found the issue, so I've removed it locally and will be removed as I start cleaning the code.

SkywingvL replied:
I see.

Useful debugging techniques for this sort of situation that I usually use, for reference:

1) Putting in an if (IsDebuggerPresent())) __debugbreak();
2) windbg -o app (or .childbg 1) so that child processes are debugged too.

SkywingvL commented on src/plugins/legacy_plugin.h:
```diff
@@ -35,7 +37,7 @@ class LegacyPlugin
 
 	// Called when a plugin DLL gets loaded.
-	virtual bool Init(char*);
+	virtual bool Init(TCHAR*);
```
These (plugin vtable methods in the plugin ABI) can't be made TCHAR.  The plugin vtable is a fixed binary standard and the pre-existing versions must not change, for backwards compatibility.  (And the legacy plugin vtable support is only there exclusively for backwards compatibility; revising it to use wchar_t instead of char in the future would defeat the whole purpose of including the legacy plugin vtable logic at all.)

scottmunday84 replied:
An artifact from old code. TCHAR was a quick solution while testing Unicode flagging. I've been starting to build wchar_t's and wstrings throughout my code.

SkywingvL commented on src/controller/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+elseif(${CMAKE_BUILD_TYPE} STREQUAL "Release")
+    target_include_directories(NWNx4Controller PRIVATE ../controller ../../lib/NWN2/include ${VCPKG_INCLUDE_DIR} ${VCPKG_LIB_DIR_RELEASE}/mswu)
+    target_link_directories(NWNx4Controller PRIVATE ../lib/NWN2 ${VCPKG_LIB_DIR_RELEASE})
+    target_link_libraries(NWNx4Controller PRIVATE ws2_32.lib wsock32.lib detours.lib wxmsw31u_core.lib wxbase31u.lib wxregexu.lib gRPC::grpc++)
```
It seems architecturally unfortunate to include gRPC in the NWNX controller.  The controller's job is purely to orchestrate starting/stopping nwn2server, monitor its responsiveness and deal with recovery from crashes, and to bootstrap the process of loading plugins into nwn2server.

gRPC seems to have a non-trivial network-accessible attack surface, and so it's reasonably likely that anything linking to gRPC will need to be updated (and all servers will have to take updates etc.) when there are inevitable security vulnerabilities found & fixed upstream in gRPC.  Ideally, it seems best to only burden environments that are going to use the gRPC-based functionality with the need to rapidly track & deploy any such fixes... can any future logic to use gRPC live, instead, in its own plugin, so that only environments that deploy that plugin need to worry about any possible gRPC issues (especially if there's listener-like code that would be involved)?

scottmunday84 replied:
This was boilerplate code to get the work running first, then isolate it and place it into another plugin. It is being moved into an xp_rpc plugin that I'm building on my end.

SkywingvL commented:
With respect to TCHAR usage in general (nonwithstanding the considerations peculiar to the plugin vtables), I'd advise *not* converting code to use TCHAR and instead just converge on either char or wchar_t.

TCHAR was intended to make code more portable between Win9x and NT, for programs that want to dual-build what was (then) ANSI (Win9x) vs. Unicode (NT) build targets.  Needless to say, nobody ought to be running any of this stuff on Win9x these days.  There's no real benefit to maintaining the dual build support, and if it's not being maintained, it's a good bet that it's probably going to quickly decay to many compile errors due to dependencies on the config being either char or wchar_t anyway.

scottmunday84 commented:
@SkywingvL In respect to TCHAR, I agree. I'm removing TCHAR, but it was used while I tested incorporating Unicode built libraries and executables alongside the new work. Legacy from my original code to get the system working. I'll be dropping TCHAR references in the coming weeks.

scottmunday84 commented:
@SkywingvL Dropped all references to TCHAR. Dropped sleep (no longer necessary in my branch). Isolating gRPC next week.

CromFr commented on src/controller/main.cpp:
```diff
@@ -18,9 +18,9 @@
 
 ***************************************************************************/
 
+#include "stdwx.h"
```
wxwidgets dependency

scottmunday84 replied:
Fixed.

CromFr commented on src/controller/service.cpp:
```diff
@@ -18,6 +18,7 @@
 
 ***************************************************************************/
 
+#include "stdwx.h"
```
wxwidgets dependency

scottmunday84 replied:
Fixed.

CromFr commented on src/controller/service.cpp:
```diff
@@ -59,10 +60,10 @@ BOOL installservice(int serviceNo)
-    snprintf(displayName, 64, "NWNX4 Service %d", serviceNo);
-    snprintf(cmdLine, MAX_PATH, "%s -serviceno %d -runservice", szPath, serviceNo);
-	sdBuf.lpDescription = "Neverwinter Nights Extender 4 service instance";
+	_stprintf_s(serviceName, 64, _T("NWNX4-%d"), serviceNo);
```
TCHAR stuff

scottmunday84 replied:
Fixed.

CromFr commented on src/controller/stdwx.cpp:
```diff
@@ -0,0 +1 @@
+#include "stdwx.h"
```
Both .cpp & .h file should be removed

scottmunday84 replied:
Fixed.

CromFr commented on src/controller/udp.cpp:
```diff
@@ -18,6 +18,7 @@
     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  ***************************************************************************/
 
+#include "stdwx.h"
```
wxwidgets

scottmunday84 replied:
Fixed.

CromFr reviewed (changes requested):
I'm really glad you're here to help me refresh nwnx4 :)

It looks like your PR reverts some changes I made to remove wxwidgets dependencies. Ideally only the GUI should depend on wxwidgets. I would suggest that you go through [the PR changes](https://github.com/nwn2dev/nwnx4/pull/2/files) and look for `wx` occurences, and also remove wxwidgets-related defines, include dirs and libraries from CMakeLists.txt files.

Can you edit the README.md file to add instructions for compiling the project using CMake (what must be installed, how to generate the makefiles, etc.) ? You can remove the VS2019 instructions if they are now obsolete (and remove .vcxproj files?).

Also, what is util/wx-config.exe for? Is it relevant to keep it versioned in the repo?

Out of curiosity, do you know what is the "modern" way to use Windows APIs ? Is it recommended for new projects to always define _UNICODE ?

SkywingvL commented:
In general, NT is natively 16-bit wchar_t.  If you're building code that only needs to care about Windows (certainly the controller and plugins are likely to fit into this category), you might as well use wchar_t.  Otherwise you'll end up wanting to convert from UTF-8 to wchar_t manually whenever you interface with platform APIs (char isn't UTF-8 on most NT platforms).

With respect to CMake, I'd consider whether it's beneficial to use CMake for platform-specific components or not.  My (very limited) experience with CMake has been that CMake is at times rather cumbersome to deal with for a project that isn't cross-platform over using a platform's de-facto build environment, especially when it comes to things like precompiled headers (which will often dramatically cut compile times), or relatively recently introduced compiler flags... but perhaps it's adding some benefits here?

scottmunday84 commented on src/controller/controller.cpp:
```diff
@@ -114,10 +115,10 @@ bool NWNXController::startServerProcessInternal()
 
-	char szDllPath[MAX_PATH];
-	char* pszFilePart = NULL;
+	TCHAR szDllPath[MAX_PATH];
```
Fixed.

scottmunday84 commented on src/controller/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+project(NWNx4Controller VERSION 1.0)
+
```
Done. It was an artifact from my branch.

scottmunday84 commented on .idea/.gitignore:
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
I think it would be nice. It gives a space to start from if people are working in a particular IDE. I use CLion and some of the additional IDE files will define the configurations of how it is to build out the project along with its attachment to the default toolchain etc.

scottmunday84 commented on src/controller/controller.cpp:
```diff
@@ -19,14 +19,13 @@
 
 NWNXController::NWNXController(SimpleIniConfig* config)
 {
-	this->config = config;
```
I'll fix this.

scottmunday84 replied:
Fixed.

scottmunday84 commented on src/controller/controller.cpp:
```diff
@@ -19,14 +19,13 @@
 
 ***************************************************************************/
 
+#include "stdwx.h"
```
Dropped.

scottmunday84 commented:
> In general, NT is natively 16-bit wchar_t. If you're building code that only needs to care about Windows (certainly the controller and plugins are likely to fit into this category), you might as well use wchar_t. Otherwise you'll end up wanting to convert from UTF-8 to wchar_t manually whenever you interface with platform APIs (char isn't UTF-8 on most NT platforms).
> 
> With respect to CMake, I'd consider whether it's beneficial to use CMake for platform-specific components or not. My (very limited) experience with CMake has been that CMake is at times rather cumbersome to deal with for a project that isn't cross-platform over using a platform's de-facto build environment, especially when it comes to things like precompiled headers (which will often dramatically cut compile times), or relatively recently introduced compiler flags... but perhaps it's adding some benefits here?

CMake typically is flexible enough to work in multiple platforms. It is one reason why JetBrains chose it as its default project file for CLion and can be read in by both VS Code and Visual Studio. I'm open for suggestions, but CMake seemed like a great middle ground. I've also seen some push toward using bazel, but it doesn't have the best of support currently. I also know @CromFr suggested Meson. I'm all for any of these choices. I used CMake on my original branch, so that's why I continued to use it here.

scottmunday84 commented:
Cleaned up more from this PR. Will develop some other ideas into another PR.

* nullptr replacement has been corrected: only persistent in gui, controller, and hooks.
* Dropped .idea files.

CromFr reviewed (approved):
LGTM !

Thanks a lot for your contribution ! Having a functional GUI is a huge step towards feature parity with existing nwnx4 releases !

I'm going to tune a bit meson to automate as much as possible the build process for CD/CI, so we can provide binaries on the Github release page.

scottmunday84 reviewed (commented):
> LGTM !
> 
> Thanks a lot for your contribution ! Having a functional GUI is a huge step towards feature parity with existing nwnx4 releases !
> 
> I'm going to tune a bit meson to automate as much as possible the build process for CD/CI, so we can provide binaries on the Github release page.

Thanks, @CromFr! I'm going to try finishing up my new PR with gRPC in the next week or two.
